runtime/lua/vim/_memoize.lua

```diff
@@ -0,0 +1,51 @@
+--- Module for private utility functions
+
+local function concat_hash(...)
+  return table.concat({ ... })
+end
+
+local function idx_hash(idx)
+  return function(...)
+    return select(idx, ...)
+  end
+end
+
+--- @param hash_fn any
+--- @return fun(...): any
+local function resolve_hash(hash_fn)
+  if type(hash_fn) == 'number' then
+    hash_fn = idx_hash(hash_fn)
+  elseif type(hash_fn) == 'string' then
+    if hash_fn == 'concat' then
+      hash_fn = concat_hash
+    else
+      error('invalid value for hash_fn: ' .. hash_fn)
+    end
+  end
+  return hash_fn
+end
+
+--- @generic F: function
+--- @param hash? integer|'concat'|fun(...): any
+--- @param fn F
+--- @return F
+return function(hash, fn)
+  vim.validate({
+    hash = { hash, { 'integer', 'string', 'function' } },
```

Comment:
```suggestion
    hash = { hash, { 'number', 'string', 'function' } },
```
`'integer'` is an invalid type name:
<img width="1011" alt="image" src="https://github.com/neovim/neovim/assets/62502207/68ee7079-68ab-40d7-97c2-f025951f71e4">